pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/LookupTransformFunction.java

```diff
@@ -0,0 +1,332 @@
+        "Second argument must be a literal(string) representing the column name from dimension table to lookup");
+    _dimColumnName = ((LiteralTransformFunction) dimColumnFunction).getLiteral();
+
+    List<TransformFunction> joinArguments = arguments.subList(2, arguments.size());
+    int numJoinArguments = joinArguments.size();
+    for (int i = 0; i < numJoinArguments / 2; i++) {
+      TransformFunction dimJoinKeyFunction = joinArguments.get((i * 2));
+      Preconditions.checkArgument(dimJoinKeyFunction instanceof LiteralTransformFunction,
+          "JoinKey argument must be a literal(string) representing the primary key for the dimension table");
+      _joinKeys.add(((LiteralTransformFunction) dimJoinKeyFunction).getLiteral());
+
+      TransformFunction factJoinValueFunction = joinArguments.get((i * 2) + 1);
+      TransformResultMetadata factJoinValueFunctionResultMetadata = factJoinValueFunction.getResultMetadata();
+      Preconditions.checkArgument(factJoinValueFunctionResultMetadata.isSingleValue(),
+          "JoinValue argument must be a single value expression");
+      _joinValueFunctions.add(factJoinValueFunction);
+    }
+
+    // Validate lookup table and relevant columns
+    _dataManager = DimensionTableDataManager.getInstanceByTableName(_dimTableName);
+    Preconditions.checkArgument(_dataManager != null,
+            String.format("Dimension table does not exist: %s", _dimTableName));
+
+    _lookupColumnFieldSpec = _dataManager.getColumnFieldSpec(_dimColumnName);
+    Preconditions.checkArgument(_lookupColumnFieldSpec != null,
+        String.format("Column does not exist in dimension table: %s:%s", _dimTableName, _dimColumnName));
+
+    for (String joinKey : _joinKeys) {
+      FieldSpec pkColumnSpec = _dataManager.getColumnFieldSpec(joinKey);
+      Preconditions.checkArgument(pkColumnSpec != null,
+          String.format("Primary key column doesn't exist in dimension table: %s:%s", _dimTableName, joinKey));
+      _joinValueFieldSpecs.add(pkColumnSpec);
+    }
+
+    List<String> tablePrimaryKeyColumns = _dataManager.getPrimaryKeyColumns();
+    Preconditions.checkArgument(_joinKeys.equals(tablePrimaryKeyColumns),
+        String.format("Provided join keys (%s) must be the same as table primary keys: %s", _joinKeys,
+            tablePrimaryKeyColumns));
+  }
+
+  @Override
+  public TransformResultMetadata getResultMetadata() {
+    return new TransformResultMetadata(_lookupColumnFieldSpec.getDataType(),
+        _lookupColumnFieldSpec.isSingleValueField(), false);
+  }
+
+  private Object[] lookup(ProjectionBlock projectionBlock) {
+    int numPkColumns = _joinKeys.size();
+    int numDocuments = projectionBlock.getNumDocs();
+    Object[][] pkColumns = new Object[numPkColumns][];
+    for (int i = 0; i < numPkColumns; i++) {
+      FieldSpec.DataType colType = _joinValueFieldSpecs.get(i).getDataType();
+      TransformFunction tf = _joinValueFunctions.get(i);
+      switch (colType) {
+        case STRING:
```

Comment:
(nit) move `STRING` after `DOUBLE` for consistency

Reply:
Moved.